unexpected indent of test function with docstring

When run test with docstring, FAILURE message indent is unexpected.

For example, 

```python
import unittest
class TestSome(unittest.TestCase):

    def test_bar(self):
        """
        This test fails.
        """
        self.assertEqual(1, 3)
```

run above test with pytest, following message is shown.

```
    def test_foo(self):
        """
            This test fails.
            """
>       self.assertEqual(1, 3)
E       AssertionError: 1 != 3
```

But I expected message like below.

```
    def test_foo(self):
        """
        This test fails.
        """
>       self.assertEqual(1, 3)
E       AssertionError: 1 != 3
```

I found out the cause is probably `tokenize.generate_tokens`. (It is used `src/_pytest/_code/source.deindent` ). However, I do not know how to fix it.

# Version

- macOS 10.14
- Python: 3.6.5
- pytest: 3.8.1

I can reproduce this as well.

While debugging we _have_ the right source:

```
(Pdb) print('\n'.join(self.frame.code.fullsource))
import unittest
class TestSome(unittest.TestCase):

    def test_bar(self):
        """
        This test fails.
        """
        self.assertEqual(1, 3)
(Pdb) print(source[start:end])
    def test_bar(self):
        """
        This test fails.
        """
        self.assertEqual(1, 3)
```

And here's where it goes awry:

```
(Pdb) n
> /home/asottile/workspace/pytest/src/_pytest/_code/code.py(540)_getentrysource()
-> source = source.deindent()
(Pdb) n
> /home/asottile/workspace/pytest/src/_pytest/_code/code.py(541)_getentrysource()
-> return source
(Pdb) print(source)
def test_bar(self):
    """
        This test fails.
        """
    self.assertEqual(1, 3)
```

This does appear intentional, but weird.  `deindent` uses the tokenizer to do indentation changes.  Since a docstring is a multi line string, the whitespace _is_ important and so the start of the token becomes deindented but not the whitespace contained.

And actually, here's a great example where the deindent actually introduces a misleading output:

```python
def test_bar():
    x = """\
    sneaky!"""
    assert x == '        sneaky!'
```

```
$ pytest t.py 
============================= test session starts ==============================
platform linux -- Python 3.6.5, pytest-3.8.3.dev1+gdf435fa8.d20181002, py-1.6.0, pluggy-0.7.1
rootdir: /tmp/t, inifile:
collected 1 item                                                               

t.py F                                                                   [100%]

=================================== FAILURES ===================================
___________________________________ test_bar ___________________________________

    def test_bar():
        x = """\
        sneaky!"""
>       assert x == '        sneaky!'
E       AssertionError: assert '    sneaky!' == '        sneaky!'
E         -     sneaky!
E         +         sneaky!
E         ? ++++
```

if I were to copy paste the code out of that:

```pycon
>>> x = """\
...         sneaky!"""
>>> assert x == '        sneaky!'
>>> 
```

it passes! wow!

my inkling is that this is one of those "damned if you do damned if you don't" situations I think we should make the indentation make sense with the source text and remove our fancy tokenize dedent logic and just use `textwrap.dedent` -- @nicoddemus @RonnyPfannschmidt thoughts?

Hehehe I was just debugging this one. 😁 

I replaced `deindent` with this:

```python
def deindent(lines, offset=None):
    import textwrap
    return textwrap.dedent('\n'.join(lines)).split('\n')
```

With this new implementation I get the expected output:

```
============================================ FAILURES =============================================
________________________________________ TestSome.test_bar ________________________________________

self = <test-dedent.TestSome instance at 0x04598490>

    def test_bar(self):
        """
        This test fails.
        """
>       assert 1 == 3
E       assert 1 == 3

test-dedent.py:8: AssertionError
==================================== 1 failed in 0.03 seconds =====================================
``` 

While the original implementation gives:

```
============================================ FAILURES =============================================
________________________________________ TestSome.test_bar ________________________________________

self = <test-dedent.TestSome instance at 0x047D5508>

    def test_bar(self):
        """
            This test fails.
            """
>       assert 1 == 3
E       assert 1 == 3

test-dedent.py:8: AssertionError
==================================== 1 failed in 0.03 seconds =====================================
```

But 4 tests in `test_source.py` fail because of that. One of the tests actually seem to be incorrect and the new output is the correct one, the others I didn't take a look yet.

Want to assume this one @asottile? I won't have much more time to work on this one anyway.

--- a/changelog/4066.bugfix.rst
+++ b/changelog/4066.bugfix.rst
@@ -1,0 +1,1 @@
+Fix source reindenting by using ``textwrap.dedent`` directly.

--- a/src/_pytest/_code/source.py
+++ b/src/_pytest/_code/source.py
@@ -7,6 +7,7 @@ import linecache
 import sys
 import six
 import inspect
+import textwrap
 import tokenize
 import py
 
@@ -23,7 +24,6 @@ class Source(object):
     def __init__(self, *parts, **kwargs):
         self.lines = lines = []
         de = kwargs.get("deindent", True)
-        rstrip = kwargs.get("rstrip", True)
         for part in parts:
             if not part:
                 partlines = []
@@ -33,11 +33,6 @@ class Source(object):
                 partlines = [x.rstrip("\n") for x in part]
             elif isinstance(part, six.string_types):
                 partlines = part.split("\n")
-                if rstrip:
-                    while partlines:
-                        if partlines[-1].strip():
-                            break
-                        partlines.pop()
             else:
                 partlines = getsource(part, deindent=de).lines
             if de:
@@ -269,46 +264,7 @@ def getsource(obj, **kwargs):
 
 
 def deindent(lines, offset=None):
-    if offset is None:
-        for line in lines:
-            line = line.expandtabs()
-            s = line.lstrip()
-            if s:
-                offset = len(line) - len(s)
-                break
-        else:
-            offset = 0
-    if offset == 0:
-        return list(lines)
-    newlines = []
-
-    def readline_generator(lines):
-        for line in lines:
-            yield line + "\n"
-
-    it = readline_generator(lines)
-
-    try:
-        for _, _, (sline, _), (eline, _), _ in tokenize.generate_tokens(
-            lambda: next(it)
-        ):
-            if sline > len(lines):
-                break  # End of input reached
-            if sline > len(newlines):
-                line = lines[sline - 1].expandtabs()
-                if line.lstrip() and line[:offset].isspace():
-                    line = line[offset:]  # Deindent
-                newlines.append(line)
-
-            for i in range(sline, eline):
-                # Don't deindent continuing lines of
-                # multiline tokens (i.e. multiline strings)
-                newlines.append(lines[i])
-    except (IndentationError, tokenize.TokenError):
-        pass
-    # Add any lines we didn't see. E.g. if an exception was raised.
-    newlines.extend(lines[len(newlines) :])
-    return newlines
+    return textwrap.dedent("\n".join(lines)).splitlines()
 
 
 def get_statement_startend2(lineno, node):

--- a/src/_pytest/_code/source.py
+++ b/src/_pytest/_code/source.py
@@ -110,17 +110,10 @@ class Source(object):
         ast, start, end = getstatementrange_ast(lineno, self)
         return start, end
 
-    def deindent(self, offset=None):
-        """ return a new source object deindented by offset.
-            If offset is None then guess an indentation offset from
-            the first non-blank line.  Subsequent lines which have a
-            lower indentation offset will be copied verbatim as
-            they are assumed to be part of multilines.
-        """
-        # XXX maybe use the tokenizer to properly handle multiline
-        #     strings etc.pp?
+    def deindent(self):
+        """return a new source object deindented."""
         newsource = Source()
-        newsource.lines[:] = deindent(self.lines, offset)
+        newsource.lines[:] = deindent(self.lines)
         return newsource
 
     def isparseable(self, deindent=True):
@@ -263,7 +256,7 @@ def getsource(obj, **kwargs):
     return Source(strsrc, **kwargs)
 
 
-def deindent(lines, offset=None):
+def deindent(lines):
     return textwrap.dedent("\n".join(lines)).splitlines()
 
 

--- a/testing/code/test_source.py
+++ b/testing/code/test_source.py
@@ -405,21 +405,13 @@ def test_deindent():
 
     assert deindent(["\tfoo", "\tbar"]) == ["foo", "bar"]
 
-    def f():
-        c = """while True:
-    pass
-"""
-
-    lines = deindent(inspect.getsource(f).splitlines())
-    assert lines == ["    def f():", '        c = """while True:', "    pass", '"""']
-
-    source = """
+    source = """\
         def f():
             def g():
                 pass
     """
     lines = deindent(source.splitlines())
-    assert lines == ["", "def f():", "    def g():", "        pass"]
+    assert lines == ["def f():", "    def g():", "        pass"]
 
 
 def test_source_of_class_at_eof_without_newline(tmpdir):

--- a/testing/code/test_source.py
+++ b/testing/code/test_source.py
@@ -27,16 +27,7 @@ def test_source_str_function():
     x = Source(
         """
         3
-    """,
-        rstrip=False,
-    )
-    assert str(x) == "\n3\n    "
-
-    x = Source(
         """
-        3
-    """,
-        rstrip=True,
     )
     assert str(x) == "\n3"
 
@@ -400,10 +391,13 @@ def test_getfuncsource_with_multine_string():
     pass
 """
 
-    assert (
-        str(_pytest._code.Source(f)).strip()
-        == 'def f():\n    c = """while True:\n    pass\n"""'
-    )
+    expected = '''\
+    def f():
+        c = """while True:
+    pass
+"""
+'''
+    assert str(_pytest._code.Source(f)) == expected.rstrip()
 
 
 def test_deindent():
@@ -417,7 +411,7 @@ def test_deindent():
 """
 
     lines = deindent(inspect.getsource(f).splitlines())
-    assert lines == ["def f():", '    c = """while True:', "    pass", '"""']
+    assert lines == ["    def f():", '        c = """while True:', "    pass", '"""']
 
     source = """
         def f():
@@ -425,7 +419,7 @@ def test_deindent():
                 pass
     """
     lines = deindent(source.splitlines())
-    assert lines == ["", "def f():", "    def g():", "        pass", "    "]
+    assert lines == ["", "def f():", "    def g():", "        pass"]
 
 
 def test_source_of_class_at_eof_without_newline(tmpdir):

